### PR TITLE
feat: directive to dismiss or close active modal

### LINKED ISSUE #9428: Close added-to-cart modal using modal directive
Currently the Added-to-cart modal needs to be closed explicitly using its dismissModal() method. It's now required when:
1. item counter goes down to 0 (item removed) OR
2. any navigation link is clicked (even in nested components like cart-item.component or its children)

But the cart-item.component and deeply-nested children have no business to know they are projected in a modal or not. Therefore they shouldn't be obligated to explicitly close a modal when clicking some link inside those components.

~Instead, we should automatically detect router navigation and close the added-to-cart modal when the router event happens.~

It's a breaking change required in 3.0. It will help get rid of unnecessary outputs (public-api) from cart-item.component, which cause problems when adding outlets inside cart-item.component for Product Configurator feature (which is likely coming in next minor releases).

### PR DESCRIPTION
**Pull request by Platonn**
Previously `CartItemComponent` had `@Output() view` and method `viewItem()` only to communicate with the parent component `AddedToCartDialogComponent` - just to close the modal. However `CartItemComponent` was rendered in many other contexts (cart details page, order summary, etc.), where the `view` Output was not used and didn't make too much sense.

Say, we want to introduce a link that should also close the modal inside some nested outlet component inside the cart item. Then we would need to peform similar bubbling of Output events upwards just to close the modal.

To avoid it, now it's simplified, thanks to the new `cxModal` directive, which on click event dismisses or closes the active modal (depending on the given `type` and `reason`). But it performs it only when the `cxModal` is rendered inside some active modal. Otherwise clicking the directive doesn't cause any effect (`cxModal` directive can be placed on any element, regardless if it's redered inside modal or not).

BREAKING CHANGE:
`AddedToCartDialogComponent` and `CartItemComponent` now depend on the `cxModal` directive. Moreover `CartItemComponent` lost its `@Output() view` and its method `viewItem()`.

closes GH-9428

### CODE REVIEW

**Review by marlass — COMMENTED** [review]
One small change in behavior noticed:
- clicking on the link to the product page when you are already on this page doesn't close modal (previously it did).

File: projects/storefrontlib/src/cms-components/cart/add-to-cart/added-to-cart-dialog/added-to-cart-dialog.component.ts [hunk 1]
```diff
@@ -83,7 +90,8 @@ export class AddedToCartDialogComponent implements OnInit {
             })
           )
         ),
-        map(() => <FormControl>this.form.get('quantity'))
+        map(() => <FormControl>this.form.get('quantity')),
+        shareReplay({ bufferSize: 1, refCount: true })
```
Comment (Platonn) [inline projects/storefrontlib/src/cms-components/cart/add-to-cart/added-to-cart-dialog/added-to-cart-dialog.component.ts:88]: Unrelated change. By the way. Just for safety.

File: projects/storefrontlib/src/shared/components/modal/modal.directive.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+   * Reason for closing/dismissing the modal. Fallbacks to URL of the `routerLink` (if exists on the same element).
+   */
+  @Input() cxModalReason: ModalDirectiveOptions['reason'];
+
+  constructor(
+    protected service: ModalDirectiveService,
+    protected router: Router,
+
+    // active modal can be injected only when the directive is projected inside modal
+    @Optional() protected activeModal: NgbActiveModal,
```
Comment (Platonn) [inline projects/storefrontlib/src/shared/components/modal/modal.directive.ts:80]: Note: depedency on `NgBootstap` library. It provides locally (for the activated component) the local service to close/dismiss the surrounding modal.

This helps to avoid closing irrelevant modal.

File: projects/storefrontlib/src/shared/components/modal/modal.directive.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+   */
+  reason?: string;
+}
+
+/**
+ * Service to control the behavior of the DismissModalDirective
+ */
+@Injectable({
+  providedIn: 'root',
+})
```
Comment (Platonn) [inline projects/storefrontlib/src/shared/components/modal/modal.directive.ts:28]: Service provided in root to allow for extensibility.
Use case: when customer doesn't want to close a modal in some exceptional case, even if the directive was applied in our component.

File: projects/storefrontlib/src/shared/components/modal/modal.directive.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    @Optional() protected activeModal: NgbActiveModal,
+
+    @Optional() protected routerLink: RouterLink,
+    @Optional() protected routerLinkWithHref: RouterLinkWithHref
+  ) {}
+
+  /**
+   * Returns URL in case when the directive is in the scope of a `routerLink` directive at the same time.
+   */
+  protected getUrl(): string {
```
Comment (marlass) [inline projects/storefrontlib/src/shared/components/modal/modal.directive.ts]: Could you also mark that this can return `undefined`? Some day we will want to use `strict` TS mode.

**Review by marlass — DISMISSED** [review]

**Review by marlass — APPROVED** [review]
